### ISSUE TITLE
Tree shaking when concatenating strings?

### ISSUE THREAD

**Issue by eigilsagafos** [evanw/esbuild#345]
Given the following structure:

src/lib/fragment.js
```js
export const fragment = "fragment"
```
src/lib/stringA.js
```js
import { fragment } from "./fragment"
export const StringA = "String A with fragment: " + fragment
```
src/lib/stringB.js
```js
export const StringB = "String B without fragment"
```
src/lib/stringC.js
```js
export const StringC = "String C without fragment"
```
src/lib/stringC.js
```js
export const StringC = "String C without fragment"
```
src/lib/index.js
```js
export { StringA } from "./stringA"
export { StringB } from "./stringB"
export { StringC } from './stringC'
```

src/index.js
```js
import { StringB } from "./lib"
console.log(StringB)
```

Running `esbuild src/index.js --bundle --platform=node` produces the following:
```js
// src/lib/fragment.js
const fragment = "fragment";

// src/lib/stringA.js
const StringA = "String A with fragment " + fragment;

// src/lib/stringB.js
const StringB = "String B without fragment";

// src/lib/stringC.js

// src/lib/index.js

// src/index.js
console.log(StringB);
```

I would expect `fagment` and `stringA` to be tree-shaken. If I make StringA a function it is. Is this expected? Any way around this?

**Comment by eigilsagafos** [comment]
Thanks @rtsao for taking a go at this. Any thoughts about this @evanw?

**Comment by evanw** [comment]
This is tricky because esbuild doesn't yet have support for this kind of complex cross-file analysis. This would indeed be something to aspire to but as #351 shows, doing this safely isn't easy. This would also need to check that the transformation doesn't change behavior due to the TDZ (temporal dead zone) of any reference involved, which is also analysis that esbuild doesn't do yet. Since this seems like a big step up in complexity, I'm not sure yet if it makes sense to include this kind of thing in esbuild. It might turn out to be too complex and hard to maintain. This requires more investigation.

I tried to come up with a way around this by using a function with a `/* @__PURE__ */` comment. I have adopted the convention used by other tools (e.g. Rollup, UglifyJS) that a function call preceded by a `/* @__PURE__ */` comment is considered to be removable by tree shaking if the result is unused. Unfortunately the convention only applies to call expressions and there's no way to annotate any other kind of expression. Still, you could in theory do this so that tree shaking could remove `StringA` if it's not used:

```js
function concat(a, b) {
  return a + b
}

export const fragment = "fragment"
export const StringA = /* @__PURE__ */ concat("String A with fragment: ", fragment)
```

However, it looks like that currently only works if the referenced variables are in the same file, so that wouldn't work in your case either. Making this kind of tree shaking work across files is also an area of future work for esbuild apparently.

Long story short: I don't think this is possible yet.

**Comment by eigilsagafos** [comment]
Thank you @evanw for a thorough answer and for taking the time to test potential workarounds. I'll close this for now.

**Comment by crimx** [comment]
FYI this also works

```js
let a = '222'
let b = /* @__PURE__ */ a.concat('222')
```
